R/functions.R

```diff
@@ -25,15 +25,23 @@ get_data <- function(synID, version = NULL){
 #'@param factors A vector of factor variables
 #'@param continuous A vector of continuous variables
 #' Remove primary variable for now - A vector of primary variables
-#'@param sample_variable A character vector with the column name corresponding to sample identifiers
 #'
 #'@export
 #'@return A data frame with coerced variables.
 #'@examples
-clean_covariates <- function(md, factors, continuous, sample_variable){
-  md[,factors] <- lapply(md[,factors], factor)
-  md[,continuous] <- lapply(md[,continuous], as.numeric)
-  md
+clean_covariates <- function(md, factors, continuous){
+  if (missing(factors) | missing(continuous)) {
+    stop("Factor and continuous variables are required.")
+  } else if (length(intersect(factors, continuous)) != 0) {
+    stop("Variables are present in both the continuous and factor arguments. Variables must be designated
+         numeric or factor, not both.")
+  } else if (!(factors %in% colnames(md) | !(continuous %in% colnames(md)))) {
+    stop("Variables provided are not present in the metadata.")
+  } else {
+    md[,factors] <- lapply(md[,factors], factor)
+    md[,continuous] <- lapply(md[,continuous], as.numeric)
```

Comment:
```suggestion
    md[,continuous] <- lapply(md[,continuous, drop = FALSE], as.numeric)
```

Reply:
Also, what is the expected behavior here when the original columns are characters vs factors? `as.numeric()` on a character column will return `NA` and a warning; on a factor column it will return the numeric level of the factor. See below:

``` r
dat <- data.frame(x = 1:3, y = c("a", "b", "c"), z = c("d", "e", "f"))
dat$y <- as.character(dat$y)
str(dat)
#> 'data.frame':    3 obs. of  3 variables:
#>  $ x: int  1 2 3
#>  $ y: chr  "a" "b" "c"
#>  $ z: Factor w/ 3 levels "d","e","f": 1 2 3
clean_covariates(dat, factors = "x", continuous = c("y", "z"))
#> Warning in if (!(factors %in% colnames(md) | !(continuous %in%
#> colnames(md)))) {: the condition has length > 1 and only the first element
#> will be used
#> Warning in lapply(md[, continuous, drop = FALSE], as.numeric): NAs
#> introduced by coercion
#>   x  y z
#> 1 1 NA 1
#> 2 2 NA 2
#> 3 3 NA 3
```

Reply:
@karawoo hm - Not totally sure what to do in this case. I could: 
1. import data with all column types as character so they are more easily coercible to factors and numeric.
AND 
2. add a message that only coerces a vector if it passes `type.convert` 
```
predict_coerce <- function(vec) {
  if (class(type.convert(vec)) == "numeric"){
  vec <- as.numeric(vec)
  vec
} else {
  message("Variable can not be coerced to numeric.")
}
}

predict_coerce(dat$y)
#>Variable can not be coerced to numeric.
```

Reply:
I didn't know about `type.convert()`, that's cool. I think that would work, though you would want to also allow `class(type.convert(vec)) %in% c("integer", "numeric")`, not only numeric.

Another option is to convert only the continuous columns to character first before converting them to numeric. It seems counterintuitive but this will ensure that character and factor columns will get converted the same way. I don't think this is any better than what you suggested, but I'll share here for thoroughness. You will still get NAs and a warning if they can't be converted to numeric:

``` r
d1 <- c("1.5", "2.1", "3,7")
d2 <- factor(c("1.5", "2.1", "3,7"))

as.numeric(as.character(d1))
#> Warning: NAs introduced by coercion
#> [1] 1.5 2.1  NA
as.numeric(as.character(d2))
#> Warning: NAs introduced by coercion
#> [1] 1.5 2.1  NA
```

<sup>Created on 2019-11-12 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

If you want to force an error rather than a warning if the columns can't be converted, you could wrap this in `tryCatch()`:

```r
tryCatch(
  as.numeric(as.character(d1)),
  warning = function(w) stop(w)
)
#> Error in doTryCatch(return(expr), name, parentenv, handler) : 
#>   NAs introduced by coercion
```